### PR TITLE
Add list_test_devices method

### PR DESCRIPTION
**Pull request by waltdisgrace**
This is a preliminary PR that will address #465.

### CODE REVIEW

**Comment by waltdisgrace** [comment]
Once @bgurney-rh tests this, the rest of the team can review this.

**Comment by jbaublitz** [comment]
Going to wait to discuss this one before reviewing and merging.

**Comment by bgurney-rh** [comment]
Trying it on my test machine with a single device-mapper device, I see a failure in core::dm::tests::sudo_test_rename (but only in that test):

```
failures:

---- core::dm::tests::sudo_test_rename stdout ----
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `2`,
 right: `1`', src/core/dm.rs:951:9
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.29/src/backtrace/libunwind.rs:88
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.29/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:47
   3: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:36
   4: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:200
   5: std::panicking::default_hook
             at src/libstd/panicking.rs:211
   6: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:477
   7: std::panicking::continue_panic_fmt
             at src/libstd/panicking.rs:384
   8: std::panicking::begin_panic_fmt
             at src/libstd/panicking.rs:339
   9: devicemapper::core::dm::tests::sudo_test_rename
             at src/core/dm.rs:951
  10: devicemapper::core::dm::tests::sudo_test_rename::{{closure}}
             at src/core/dm.rs:935
  11: core::ops::function::FnOnce::call_once
             at /rustc/eae3437dfe991621e8afdc82734f4a172d7ddf9b/src/libcore/ops/function.rs:231
  12: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
             at /rustc/eae3437dfe991621e8afdc82734f4a172d7ddf9b/src/liballoc/boxed.rs:746
  13: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:82
  14: std::panicking::try
             at /rustc/eae3437dfe991621e8afdc82734f4a172d7ddf9b/src/libstd/panicking.rs:275
  15: std::panic::catch_unwind
             at /rustc/eae3437dfe991621e8afdc82734f4a172d7ddf9b/src/libstd/panic.rs:394
  16: test::run_test::run_test_inner::{{closure}}
             at src/libtest/lib.rs:1466
  17: test::run_test::run_test_inner
             at src/libtest/lib.rs:1488
  18: test::run_test
             at src/libtest/lib.rs:1507
  19: test::run_tests
             at src/libtest/lib.rs:1166
  20: test::run_tests_console
             at src/libtest/lib.rs:971
  21: test::test_main
             at src/libtest/lib.rs:292
  22: test::test_main_static
             at src/libtest/lib.rs:326
  23: devicemapper::main
  24: std::rt::lang_start::{{closure}}
             at /rustc/eae3437dfe991621e8afdc82734f4a172d7ddf9b/src/libstd/rt.rs:64
  25: std::rt::lang_start_internal::{{closure}}
             at src/libstd/rt.rs:49
  26: std::panicking::try::do_call
             at src/libstd/panicking.rs:296
  27: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:82
  28: std::panicking::try
             at src/libstd/panicking.rs:275
  29: std::panic::catch_unwind
             at src/libstd/panic.rs:394
  30: std::rt::lang_start_internal
             at src/libstd/rt.rs:48
  31: std::rt::lang_start
             at /rustc/eae3437dfe991621e8afdc82734f4a172d7ddf9b/src/libstd/rt.rs:64
  32: main
  33: __libc_start_main
  34: _start
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.


failures:
    core::dm::tests::sudo_test_rename

test result: FAILED. 69 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out
```

The affected line is here:
```
let devices = dm.list_devices().unwrap();
assert_eq!(devices.len(), 1);
```

So it looks like the test assumes there will be only one device, but that may not be true anymore.

File: src/core/dm.rs [hunk 1]
```diff
@@ -269,6 +273,16 @@ impl DM {
         Ok(devs)
     }
 
+    /// Returns a list of tuples containing DM test device names, a Device, which
+    /// holds their major and minor device numbers, and on kernels that
+    /// support it, each device's last event_nr.
+    #[cfg(test)]
+    pub fn list_test_devices(&self) -> DmResult<Vec<(DmNameBuf, Device, Option<u32>)>> {
+        let mut test_devs = self.list_devices().unwrap();
```
Comment (jbaublitz) [inline src/core/dm.rs]: While this is a testing method, it's probably better to keep any `unwrap` calls in the tests themselves. Here, I'd use the `?` operator to return early with an error instead. The tests can handle failing if an error is returned.

**Comment by jbaublitz** [comment]
@waltdisgrace If you need a walkthrough of how to define it separately in the test submodule, I can help you with that.

File: src/core/dm.rs [hunk 2]
```diff
@@ -724,6 +727,17 @@ mod tests {
 
     use super::*;
 
+    impl DM {
+        /// Returns a list of tuples containing DM test device names, a Device, which
+        /// holds their major and minor device numbers, and on kernels that
+        /// support it, each device's last event_nr.
+        pub fn list_test_devices(&self) -> DmResult<Vec<(DmNameBuf, Device, Option<u32>)>> {
```
Comment (mulkieran) [inline src/core/dm.rs]: @waltdisgrace Whoops, by _the_ test submodule I meant the one that already exists, ```test_lib.rs```. You notice that you use the method in that module. I'ld prefer it to be  _defined_ in that module, also.

**Comment by mulkieran** [comment]
@waltdisgrace I'ld like #520 to precede this PR in the commit order. #520 tries for better encapsulation, and that should affect slightly how you export things from the testing module.

File: src/core/dm.rs [hunk 3]
```diff
@@ -737,20 +737,22 @@ mod tests {
     }
 
     #[test]
-    /// Verify that if no devices have been created the list is empty.
+    /// Verify that if no devices have been created the list of test devices
+    /// is empty.
     fn sudo_test_list_devices_empty() {
-        assert!(DM::new().unwrap().list_devices().unwrap().is_empty());
+        let dm = DM::new().unwrap();
```
Comment (mulkieran) [inline src/core/dm.rs]: Please leave this as one line, not two.

File: src/testing/test_lib.rs [hunk 4]
```diff
@@ -59,7 +70,7 @@ fn get_dm() -> &'static DM {
 
 /// String that is to be concatenated with test supplied name to identify
 /// devices and filesystems generated by tests.
-static DM_TEST_ID: &str = "_dm-rs_test_delme";
+pub static DM_TEST_ID: &str = "_dm-rs_test_delme";
```
Comment (mulkieran) [inline src/testing/test_lib.rs]: I don't think you need to export ```DM_TEST_ID``` anymore, now that you've moved the ```list_test_devcies``` method into ```test_lib```.

**Pull request closed without merging**